### PR TITLE
ARROW-7011: [C++] Implement casts from float/double to decimal

### PR DESCRIPTION
**Pull request by pitrou**
Also naturally available in Python using the Array.cast() method.

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc [hunk 1]
```diff
@@ -467,6 +467,51 @@ struct CastFunctor<Decimal128Type, Decimal128Type> {
...
+    } else {
+      return *std::move(result);
+    }
+  }
+
+  int32_t out_scale_, out_precision_;
+};
+
+using SafeRealToDecimal = RealToDecimal<true>;
+using UnsafeRealToDecimal = RealToDecimal<false>;
```
Comment (wesm) [inline cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc:492]: I'm not sure the performance benefit of pruning the `AllowTruncate` branch merits the 2x code size -- what do you think about making this flag just a member of the functor?

**Review by wesm — COMMENTED** [review]
This looks very reasonable to me, glad to see the conversion functions added. I had one question about generating multiple applicators based on the allow-truncation flag, absent benchmarks showing that it's makes a meaningful difference I'll quickly switch that to be a member variable and so generate less code

File: cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc:492]: Hum, yes, sounds fine. I initially thought the code would be more different, but it doesn't really make sense to generate two codepaths now.

**Comment by wesm** [comment]
Ugh there's that pesky flake again https://github.com/apache/arrow/pull/7612/checks?check_run_id=832227898

@kszucs did we determine whether it's feasible to get backtraces on macOS in GitHub Actions?

**Comment by wesm** [comment]
+1

**Pull request closed without merging**